### PR TITLE
Onboarding to ArPow (arcade-powered source-build)

### PR DESCRIPTION
**Pull request by MichaelSimons**
This PR does the following:

1. Adds the local build infrastructure that lets ArPow (arcade-powered source-build) run in this repo. See <https://github.com/dotnet/source-build/blob/master/Documentation/planning/arcade-powered-source-build/onboarding/local-onboarding.md> for more details about how it works.

    To try it out locally, run this on Linux: `./build.sh -c Release --restore --build --pack /p:ArcadeBuildFromSource=true -bl`

2. Implements source-build CI.

    To make sure ArPow (arcade-powered source-build) keeps working in this repo, we need to add it to PR validation. We also need it to run in the official build to publish source-built artifacts that can be tested downstream.

    See <https://github.com/dotnet/source-build/blob/master/Documentation/planning/arcade-powered-source-build/onboarding/ci-onboarding.md> for ArPow CI onboarding info.

3.  Incorporates the existing [source-build patches](https://github.com/dotnet/source-build/tree/release/5.0/patches/msbuild) into the repo.

    Some background on source-build patches, for anyone who isn't familiar with previous pushes for patch incorporation:

    A patch is essentially just a commit that has been extracted from Git into a `.patch` file that can be applied on demand. The effort to build .NET from source involves creating patches because repos make changes that are incompatible with source-build and need to be fixed up after the original released source code has already been finalized. When the original repo gets PRs over time for servicing, the PR changes sometimes conflict with the source-build patches, just like a merge conflict. The patch files need to be fixed up when this happens, which is a significant maintenance problem for the source-build team.

    Several times, the source-build team has pushed for "patch incorporation". This means to merge the commit represented in the `.patch` file into the original repo's official branch. Doing so prevents patch merge conflicts, because there's no longer a patch to merge against. However, patches inevitably pile up again when getting subsequent servicing releases to work in source-build.

    ArPow lets us end this maintenance-heavy process. By running source-build inside CI, patch merge conflicts will immediately block PR validation, so fixup can be handled in place, not solely by the source-build team. Running source-build in CI also means creating new patches won't be necessary except in exceptional circumstances.

See <https://github.com/dotnet/source-build/blob/master/Documentation/planning/arcade-powered-source-build/implementation-plan.md> for more details on the ArPow implementation plan.

Fixes: https://github.com/dotnet/source-build/issues/2068.

### CODE REVIEW

**Review by BenVillalobos — COMMENTED** [review]
Without digging too much into it, running your repro command in WSL failed for me:
```
bevillal@DESKTOP-OB2IDE6:/mnt/c/src/linux/msbuild$ ./build.sh -c Release --restore --build --pack /p:ArcadeBuildFromSource=true -bl
Downloading 'https://dot.net/v1/dotnet-install.sh'
dotnet-install: Note that the intended use of this script is for Continuous Integration (CI) scenarios, where:
dotnet-install: - The SDK needs to be installed without user interaction and without admin rights.
dotnet-install: - The SDK installation doesn't need to persist across multiple CI runs.
dotnet-install: To set up a development environment or to run apps, use installers rather than this script. Visit https://dotnet.microsoft.com/download to get the installer.

dotnet-install: Downloading primary link https://dotnetcli.azureedge.net/dotnet/Sdk/6.0.100-preview.2.21155.3/dotnet-sdk-6.0.100-preview.2.21155.3-linux-x64.tar.gz
dotnet-install: Extracting zip from https://dotnetcli.azureedge.net/dotnet/Sdk/6.0.100-preview.2.21155.3/dotnet-sdk-6.0.100-preview.2.21155.3-linux-x64.tar.gz
dotnet-install: Adding to current process PATH: `/mnt/c/src/linux/msbuild/.dotnet`. Note: This change will be visible only when sourcing script.
dotnet-install: Note that the script does not resolve dependencies during installation.
dotnet-install: To check the list of dependencies, go to https://docs.microsoft.com/dotnet/core/install, select your operating system and check the "Dependencies" section.
dotnet-install: Installation finished successfully.
Some command line switches were read from the auto-response file "MSBuild.rsp". To disable this file, use the "-noAutoResponse" switch.
/mnt/c/src/linux/msbuild/.dotnet/sdk/6.0.100-preview.2.21155.3/MSBuild.dll /nologo -distributedlogger:Microsoft.DotNet.Tools.MSBuild.MSBuildLogger,/mnt/c/src/linux/msbuild/.dotnet/sdk/6.0.100-preview.2.21155.3/dotnet.dll*Microsoft.DotNet.Tools.MSBuild.MSBuildForwardingLogger,/mnt/c/src/linux/msbuild/.dotnet/sdk/6.0.100-preview.2.21155.3/dotnet.dll -maxcpucount /m -verbosity:m /v:minimal /bl:/mnt/c/src/linux/msbuild/artifacts/log/Release/ToolsetRestore.binlog /clp:Summary /clp:ErrorsOnly;NoSummary /nr:true /p:TreatWarningsAsErrors=true /p:ContinuousIntegrationBuild=false /p:__ToolsetLocationOutputFile=/mnt/c/src/linux/msbuild/artifacts/toolset/6.0.0-beta.21227.1.txt /t:__WriteToolsetLocation /warnaserror /mnt/c/src/linux/msbuild/artifacts/toolset/restore.proj
/mnt/c/src/linux/msbuild/.dotnet/sdk/6.0.100-preview.2.21155.3/MSBuild.dll /nologo -distributedlogger:Microsoft.DotNet.Tools.MSBuild.MSBuildLogger,/mnt/c/src/linux/msbuild/.dotnet/sdk/6.0.100-preview.2.21155.3/dotnet.dll*Microsoft.DotNet.Tools.MSBuild.MSBuildForwardingLogger,/mnt/c/src/linux/msbuild/.dotnet/sdk/6.0.100-preview.2.21155.3/dotnet.dll -maxcpucount /m -verbosity:m /v:minimal /bl:/mnt/c/src/linux/msbuild/artifacts/log/Release/Build.binlog /clp:Summary /nr:true /p:TreatWarningsAsErrors=true /p:ContinuousIntegrationBuild=false /p:Configuration=Release /p:RepoRoot=/mnt/c/src/linux/msbuild /p:Restore=true /p:Build=true /p:Rebuild=false /p:Test=false /p:Pack=true /p:IntegrationTest=false /p:PerformanceTest=false /p:Sign=false /p:Publish=false /p:ArcadeBuildFromSource=true /warnaserror /home/bevillal/.nuget/packages/microsoft.dotnet.arcade.sdk/6.0.0-beta.21227.1/tools/Build.proj
  Cloning repository at: /mnt/c/src/linux/msbuild/ -> /mnt/c/src/linux/msbuild/artifacts/source-build/self/src/ ...
/home/bevillal/.nuget/packages/microsoft.dotnet.arcade.sdk/6.0.0-beta.21227.1/tools/SourceBuild/SourceBuildArcadeBuild.targets(123,5): error MSB3073: The command "/home/bevillal/.nuget/packages/microsoft.dotnet.arcade.sdk/6.0.0-beta.21227.1/tools/SourceBuild/git-clone-to-dir.sh  --source "/mnt/c/src/linux/msbuild/" --dest "/mnt/c/src/linux/msbuild/artifacts/source-build/self/src/" --copy-wip --clean" exited with code 1. [/home/bevillal/.nuget/packages/microsoft.dotnet.arcade.sdk/6.0.0-beta.21227.1/tools/Build.proj]

Build FAILED.

/home/bevillal/.nuget/packages/microsoft.dotnet.arcade.sdk/6.0.0-beta.21227.1/tools/SourceBuild/SourceBuildArcadeBuild.targets(123,5): error MSB3073: The command "/home/bevillal/.nuget/packages/microsoft.dotnet.arcade.sdk/6.0.0-beta.21227.1/tools/SourceBuild/git-clone-to-dir.sh  --source "/mnt/c/src/linux/msbuild/" --dest "/mnt/c/src/linux/msbuild/artifacts/source-build/self/src/" --copy-wip --clean" exited with code 1. [/home/bevillal/.nuget/packages/microsoft.dotnet.arcade.sdk/6.0.0-beta.21227.1/tools/Build.proj]
    0 Warning(s)
    1 Error(s)
```

Say something needs to be updated for ArPow in the future, are we expected to manage those updates or will there be some darc subscription that updates version numbers?

File: eng/SourceBuild.props
Comment (MichaelSimons) [reply eng/SourceBuild.props:9]: Can you clarify which file you are referring to?  The solution filter is checked in - https://github.com/dotnet/msbuild/blob/main/MSBuild.SourceBuild.slnf

File: eng/SourceBuild.props
Comment (MichaelSimons) [reply eng/SourceBuild.props:7]: Correct, ConfigureInnerBuildArgs is not required.

File: eng/Version.Details.xml
Comment (MichaelSimons) [reply eng/Version.Details.xml:7]: We will be adorning more dependencies with the `SourceBuild` attribute but not all upstream dependencies have their SourceBuild intermediate NuGet packages created yet or msbuild is not on a version that is source built yet.  Once all repos are on ArPow we will make the appropriate updates here to eliminate prebuilts.  We will also be turning on prebuilt detection in the source build CI leg to prevent regressions.

**Comment by MichaelSimons** [comment]
> Say something needs to be updated for ArPow in the future, are we expected to manage those updates or will there be some darc subscription that updates version numbers?

Any version updates will happen via darc.  There may infrastructure changes that the source-build team will make or ask the source-build champs to help out with depending on the nature of the changes.

**Comment by MichaelSimons** [comment]
>Without digging too much into it, running your repro command in WSL failed for me:

Thanks for pointing this out.  I will take a look at this.  I have not tried this command in WSL before.

**Comment by rainersigwald** [comment]
> > Without digging too much into it, running your repro command in WSL failed for me:
> 
> Thanks for pointing this out. I will take a look at this. I have not tried this command in WSL before.

One possible complicating factor here: looks like @BenVillalobos is using a Windows checkout but building in Linux (path is `/mnt/c/src/linux/msbuild/`). That often causes problems due to filesystem case sensitivity or git doing CRLF stuff. @BenVillalobos, can you try the same somewhere that's in the Linux filesystem?

**Review by dseefeld — APPROVED** [review]

**Comment by BenVillalobos** [comment]
@rainersigwald Look like I get the same issue from `/home/`:

```
bevillal@DESKTOP-OB2IDE6:~/src/msbuild$ ./build.sh -c Release --restore --build --pack /p:ArcadeBuildFromSource=true -blDownloading 'https://dot.net/v1/dotnet-install.sh'
dotnet-install: Note that the intended use of this script is for Continuous Integration (CI) scenarios, where:
dotnet-install: - The SDK needs to be installed without user interaction and without admin rights.
dotnet-install: - The SDK installation doesn't need to persist across multiple CI runs.
dotnet-install: To set up a development environment or to run apps, use installers rather than this script. Visit https://dotnet.microsoft.com/download to get the installer.

dotnet-install: Downloading primary link https://dotnetcli.azureedge.net/dotnet/Sdk/6.0.100-preview.2.21155.3/dotnet-sdk-6.0.100-preview.2.21155.3-linux-x64.tar.gz
dotnet-install: Extracting zip from https://dotnetcli.azureedge.net/dotnet/Sdk/6.0.100-preview.2.21155.3/dotnet-sdk-6.0.100-preview.2.21155.3-linux-x64.tar.gz
dotnet-install: Adding to current process PATH: `/home/bevillal/src/msbuild/.dotnet`. Note: This change will be visible only when sourcing script.
dotnet-install: Note that the script does not resolve dependencies during installation.
dotnet-install: To check the list of dependencies, go to https://docs.microsoft.com/dotnet/core/install, select your operating system and check the "Dependencies" section.
dotnet-install: Installation finished successfully.
Some command line switches were read from the auto-response file "MSBuild.rsp". To disable this file, use the "-noAutoResponse" switch.
/home/bevillal/src/msbuild/.dotnet/sdk/6.0.100-preview.2.21155.3/MSBuild.dll /nologo -distributedlogger:Microsoft.DotNet.Tools.MSBuild.MSBuildLogger,/home/bevillal/src/msbuild/.dotnet/sdk/6.0.100-preview.2.21155.3/dotnet.dll*Microsoft.DotNet.Tools.MSBuild.MSBuildForwardingLogger,/home/bevillal/src/msbuild/.dotnet/sdk/6.0.100-preview.2.21155.3/dotnet.dll -maxcpucount /m -verbosity:m /v:minimal /bl:/home/bevillal/src/msbuild/artifacts/log/Release/ToolsetRestore.binlog /clp:Summary /clp:ErrorsOnly;NoSummary /nr:true /p:TreatWarningsAsErrors=true /p:ContinuousIntegrationBuild=false /p:__ToolsetLocationOutputFile=/home/bevillal/src/msbuild/artifacts/toolset/6.0.0-beta.21227.1.txt /t:__WriteToolsetLocation /warnaserror /home/bevillal/src/msbuild/artifacts/toolset/restore.proj
/home/bevillal/src/msbuild/.dotnet/sdk/6.0.100-preview.2.21155.3/MSBuild.dll /nologo -distributedlogger:Microsoft.DotNet.Tools.MSBuild.MSBuildLogger,/home/bevillal/src/msbuild/.dotnet/sdk/6.0.100-preview.2.21155.3/dotnet.dll*Microsoft.DotNet.Tools.MSBuild.MSBuildForwardingLogger,/home/bevillal/src/msbuild/.dotnet/sdk/6.0.100-preview.2.21155.3/dotnet.dll -maxcpucount /m -verbosity:m /v:minimal /bl:/home/bevillal/src/msbuild/artifacts/log/Release/Build.binlog /clp:Summary /nr:true /p:TreatWarningsAsErrors=true /p:ContinuousIntegrationBuild=false /p:Configuration=Release /p:RepoRoot=/home/bevillal/src/msbuild /p:Restore=true /p:Build=true /p:Rebuild=false /p:Test=false /p:Pack=true /p:IntegrationTest=false /p:PerformanceTest=false /p:Sign=false /p:Publish=false /p:ArcadeBuildFromSource=true /warnaserror /home/bevillal/.nuget/packages/microsoft.dotnet.arcade.sdk/6.0.0-beta.21227.1/tools/Build.proj
  Cloning repository at: /home/bevillal/src/msbuild/ -> /home/bevillal/src/msbuild/artifacts/source-build/self/src/ ...
/home/bevillal/.nuget/packages/microsoft.dotnet.arcade.sdk/6.0.0-beta.21227.1/tools/SourceBuild/SourceBuildArcadeBuild.targets(123,5): error MSB3073: The command "/home/bevillal/.nuget/packages/microsoft.dotnet.arcade.sdk/6.0.0-beta.21227.1/tools/SourceBuild/git-clone-to-dir.sh  --source "/home/bevillal/src/msbuild/" --dest "/home/bevillal/src/msbuild/artifacts/source-build/self/src/" --copy-wip --clean" exited with code 1. [/home/bevillal/.nuget/packages/microsoft.dotnet.arcade.sdk/6.0.0-beta.21227.1/tools/Build.proj]

Build FAILED.

/home/bevillal/.nuget/packages/microsoft.dotnet.arcade.sdk/6.0.0-beta.21227.1/tools/SourceBuild/SourceBuildArcadeBuild.targets(123,5): error MSB3073: The command "/home/bevillal/.nuget/packages/microsoft.dotnet.arcade.sdk/6.0.0-beta.21227.1/tools/SourceBuild/git-clone-to-dir.sh  --source "/home/bevillal/src/msbuild/" --dest "/home/bevillal/src/msbuild/artifacts/source-build/self/src/" --copy-wip --clean" exited with code 1. [/home/bevillal/.nuget/packages/microsoft.dotnet.arcade.sdk/6.0.0-beta.21227.1/tools/Build.proj]
    0 Warning(s)
    1 Error(s)

Time Elapsed 00:00:01.19
Build failed with exit code 1. Check errors above.
```

**Comment by MichaelSimons** [comment]
I was able to build without issue in my WSL environment (Debian Buster).  I built in a fresh clone created in `/home`.  @BenVillalobos, what distro are you using?

File: eng/Packages.props
Comment (rainersigwald) [inline eng/Packages.props:28]: Why this change? It makes me nervous because we have to keep this in sync with the VS repo.

**Review by rainersigwald — CHANGES_REQUESTED** [review]

File: .vsts-dotnet.yml
Comment (MichaelSimons) [reply .vsts-dotnet.yml:236]: Yes, we need to publish the [source-build intermediate NuGet package](https://github.com/dotnet/source-build/blob/main/Documentation/planning/arcade-powered-source-build/intermediate-nupkg.md) for down stream repos' source-build legs to consume.

File: eng/Packages.props
Comment (MichaelSimons) [reply eng/Packages.props:28]: This changes allows source-build to dynamically inject the version it builds to eliminate prebuilts.  This is only done when doing a full source-build build.

**Comment by BenVillalobos** [comment]
@MichaelSimons `Ubuntu 20.04.1 LTS (GNU/Linux 4.4.0-19041-Microsoft x86_64)` from a clean clone / checkout of this branch.

**Comment by MichaelSimons** [comment]
@BenVillalobos, I was able to reproduce it with a WSL Ubuntu environment.  I opened https://github.com/dotnet/source-build/issues/2180 to track this issue.  Do you see this as a blocker for merging this PR?

File: eng/Version.Details.xml
Comment (rainersigwald) [reply eng/Version.Details.xml:7]: Note that of the list, we should only need the Roslyn one. The others are dependencies used to run tests/build a test environment that shouldn't be relevant in sourcebuild.

File: .vsts-dotnet-ci.yml
Comment (rainersigwald) [inline .vsts-dotnet-ci.yml:256]: This seems very specific; is there a mechanism to keep it updated?

**Review by rainersigwald — APPROVED** [review]

**Comment by BenVillalobos** [comment]
@MichaelSimons So long as it runs on whatever machine is generating these source builds and there's an issue tracking it (thanks for that), I don't consider it blocking

**Comment by rainersigwald** [comment]
Oh wait, have we test run the official build change? That's my only blocking issue.

**Comment by MichaelSimons** [comment]
>Oh wait, have we test run the official build change? That's my only blocking issue.

I have not.  Can you tell me if there are any special things to know when queuing?

**Comment by BenVillalobos** [comment]
Pushed up to `exp/michaelsimons/ArPow`. Should show up here soon: https://dev.azure.com/devdiv/DevDiv/_build?definitionId=9434

Side note: +1 for a bot that looks at a particular label and creates the exp/ branch for us.

**Comment by BenVillalobos** [comment]
This is the build to watch: https://dev.azure.com/devdiv/DevDiv/_build/results?buildId=4800159&view=results

File: .vsts-dotnet-ci.yml
Comment (MichaelSimons) [reply .vsts-dotnet-ci.yml:256]: I logged an issue to refactor the source-build templates to provide this as a default - https://github.com/dotnet/source-build/issues/2182

**Comment by MichaelSimons** [comment]
The test run of the official build failed. It will require a new Arcade version w/this fix - https://github.com/dotnet/arcade/pull/7453

**Comment by MichaelSimons** [comment]
@rainersigwald - thanks for the catch on running a test build!

**Comment by MichaelSimons** [comment]
To fix the issue with the official builds, we need Arcade 6.0.0-beta.21278.2 or newer.  See https://github.com/dotnet/msbuild/pull/6478.

**Comment by rainersigwald** [comment]
@Forgind can you please prioritize that?

**Comment by Forgind** [comment]
I made the change to #6478, and it passed 2 legs so far. It looks like RPS failed over the weekend, so I'll have to figure that out before we can actually merge it.

**Comment by Forgind** [comment]
I merged #6478 and fixed the merge conflict here. Should be good to merge if it passes tests, right? (I haven't actually been following this PR.)

**Comment by rainersigwald** [comment]
@Forgind let's also spin an official build by pushing the merge to `exp/michaelsimons/ArPow` too please

**Comment by MichaelSimons** [comment]
> @Forgind let's also spin an official build by pushing the merge to exp/michaelsimons/ArPow too please

I pushed it.  The new test build is https://dev.azure.com/devdiv/DevDiv/_build/results?buildId=4822514.

**Review by rainersigwald — APPROVED** [review]
The official build failure on the record parameter name thing is not blocking. I'm happy.

**Comment by MichaelSimons** [comment]
PR validation is green and according to @rainersigwald the failure in the test official build is not blocking.  Can this be merged now?

**Pull request merged**
